### DISCUSSION TITLE
JavaScriptParser class  missing?

### DISCUSSION THREAD

**Discussion by JMCTO** [antlr/grammars-v4#3615] [General]
Hi all, I'm trying to build a Javascript parser in C# VS 2019 

At a high level, I see that there is code in the C# directory which does some work. Do I still need to process the .g4 files into code?\

Also, This class is missing: 

using static JavaScriptParser;

where can I find that reference ?

I've got antlr4.runtime installed via nuget

thanks

John Mott
CTO Surgepays
jmott@surgepays.com

**Comment by kaby76** [comment]
The [javascript/javascript](https://github.com/antlr/grammars-v4/tree/master/javascript/javascript) parser for the CSharp target works. But, all that is given is the grammar ([JavaScriptLexer.g4](https://github.com/antlr/grammars-v4/blob/master/javascript/javascript/JavaScriptLexer.g4), [JavaScriptParser.g4](https://github.com/antlr/grammars-v4/blob/master/javascript/javascript/JavaScriptParser.g4)) and some support code for the parser ([JavaScriptLexerBase.cs](https://github.com/antlr/grammars-v4/blob/master/javascript/javascript/CSharp/JavaScriptLexerBase.cs), [JavaScriptParserBase.cs](https://github.com/antlr/grammars-v4/blob/master/javascript/javascript/CSharp/JavaScriptParserBase.cs)). A driver is required. And, the grammar has to be processed by the Antlr tool to generate JavaScriptParser. There is a lexer grammar, and the tool must process that as well to generate the JavaScriptLexer class, which is used by the parser.

You can write the driver code yourself, but I find that laborious. I recommend that you install my Trash toolkit, and generate a driver program. After working with that, you can fit the parser in where you need.

### Requirements
* Git. https://git-scm.com/downloads
* Install dotnet 7.x. https://dotnet.microsoft.com/en-us/download
* Install Python3. https://www.python.org/downloads/
* Install pip. https://pip.pypa.io/en/stable/installation/
* Install antlr4-tools. https://github.com/antlr/antlr4-tools
* Install Trash. https://github.com/kaby76/Domemtech.Trash#install
* ***Do not have java.exe on your path, or if you do, it must be version 11 or better! The Antlr tool is a Java jar that requires at least version 11. Do not have JAVA_HOME environmental variable set. These tend to confuse things.***

### Test requirements

Make sure all is working. Check.
```
$ python3 --version
Python 3.10.4
07/15-04:16:03 ~/issues/g4-3612/javascript/javascript/Generated-CSharp
$ pip --version
pip 23.1.2 from C:\Users\Kenne\AppData\Local\Programs\Python\Python310\lib\site-packages\pip (python 3.10)
07/15-04:16:11 ~/issues/g4-3612/javascript/javascript/Generated-CSharp
$ dotnet --version
7.0.302
07/15-04:16:17 ~/issues/g4-3612/javascript/javascript/Generated-CSharp
$ java --version
openjdk 11.0.1 2018-10-16
OpenJDK Runtime Environment 18.9 (build 11.0.1+13)
OpenJDK 64-Bit Server VM 18.9 (build 11.0.1+13, mixed mode)
$ antlr4
ANTLR Parser Generator  Version 4.13.0
 -o ___              specify output directory where all output is generated
 -lib ___            specify location of grammars, tokens files
 -atn                generate rule augmented transition network diagrams
 -encoding ___       specify grammar file encoding; e.g., euc-jp
 -message-format ___ specify output style for messages in antlr, gnu, vs2005
 -long-messages      show exception details when available for errors and warnings
 -listener           generate parse tree listener (default)
 -no-listener        don't generate parse tree listener
 -visitor            generate parse tree visitor
 -no-visitor         don't generate parse tree visitor (default)
 -package ___        specify a package/namespace for the generated code
 -depend             generate file dependencies
 -D<option>=value    set/override a grammar-level option
 -Werror             treat warnings as errors
 -XdbgST             launch StringTemplate visualizer on generated code
 -XdbgSTWait         wait for STViz to close before continuing
 -Xforce-atn         use the ATN simulator for all predictions
 -Xlog               dump lots of logging info to antlr-timestamp.log
 -Xexact-output-dir  all output goes into -o dir regardless of paths/package
07/15-04:16:58 ~/issues/g4-3612/javascript/javascript/Generated-CSharp
```

### Generate a driver
Then, in a command-line shell type the following commands. I generally use Msys (Cygwin) since cmd.exe is such a terrible shell. You can use pwsh.exe (Powershell) if you prefer.
```
git clone https://github.com/antlr/grammars-v4.git
cd grammars-v4/javascript/javascript
dotnet tool restore
dotnet trgen -- -t CSharp
cd Generated-CSharp
dotnet build
```

You can then open the .csproj in Visual Studio, then open Test.cs and read how it works.

I would recommend that you use an up-to-date version of Visual Studio. VS2019 is many years old. And, as far as I know, there is no extension for VS2019 or VS2022 that supports Antlr.

### Run
```
./bin/Debug/net7.0/Test.exe input-file.js
```

### Notes

***Do not reference [antlr4.runtime](https://www.nuget.org/packages/Antlr4.Runtime). That is OLD OLD OLD code. Remove that package reference. Read the .csproj file generated. It uses [antlr4.runtime.standard](https://www.nuget.org/packages/Antlr4.Runtime.Standard), the current, official, supported Antlr4 code. For dotnet/msbuild support of the .g4's (ie, processing the .g4's by the tool to generate .cs recognizers), use [Antlr4BuildTasks](https://www.nuget.org/packages/Antlr4BuildTasks).***

Attached is a zip file for the parser. Just type "dotnet build" to rebuild.

[Generated-CSharp.zip](https://github.com/antlr/grammars-v4/files/12063724/Generated-CSharp.zip)
